utils/gen_onnx_mlir.py

```diff
@@ -279,6 +279,38 @@
         resultTypes.push_back(attr.getType());
       } else if (auto attr = sparse_valueAttr()) {
         resultTypes.push_back(attr.getType());
+      }''',
+  "Cast":
+    '''auto toAttr = to().getSExtValue();
+      auto builder = mlir::Builder(getContext());
+      switch (toAttr) {
+        case /*onnx::TensorProto_DataType::TensorProto_DataType_FLOAT16*/10:
+          resultTypes.push_back(mlir::UnrankedTensorType::get(builder.getF16Type()));
+        case /*onnx::TensorProto_DataType::TensorProto_DataType_FLOAT*/1:
+          resultTypes.push_back(mlir::UnrankedTensorType::get(builder.getF32Type()));
+        case /*onnx::TensorProto_DataType::TensorProto_DataType_DOUBLE*/11:
+          resultTypes.push_back(mlir::UnrankedTensorType::get(builder.getF64Type()));
+        case /*onnx::TensorProto_DataType::TensorProto_DataType_INT8*/3:
+        case /*onnx::TensorProto_DataType::TensorProto_DataType_UINT8*/2:
+          resultTypes.push_back(mlir::UnrankedTensorType::get(builder.getIntegerType(/*width=*/8)));
+        case /*onnx::TensorProto_DataType::TensorProto_DataType_INT16*/5:
+        case /*onnx::TensorProto_DataType::TensorProto_DataType_UINT16*/4:
+          resultTypes.push_back(mlir::UnrankedTensorType::get(builder.getIntegerType(/*width=*/16)));
+        case /*onnx::TensorProto_DataType::TensorProto_DataType_INT32*/6:
+        case /*onnx::TensorProto_DataType::TensorProto_DataType_UINT32*/12:
+          resultTypes.push_back(mlir::UnrankedTensorType::get(builder.getIntegerType(/*width=*/32)));
+        case /*onnx::TensorProto_DataType::TensorProto_DataType_INT64*/7:
+        case /*onnx::TensorProto_DataType::TensorProto_DataType_UINT64*/13:
+          resultTypes.push_back(mlir::UnrankedTensorType::get(builder.getIntegerType(/*width=*/64)));
+        case /*onnx::TensorProto_DataType::TensorProto_DataType_BOOL*/9:
+          resultTypes.push_back(mlir::UnrankedTensorType::get(builder.getI1Type()));
+
+        case /*onnx::TensorProto_DataType::TensorProto_DataType_STRING*/8:
+        case /*onnx::TensorProto_DataType::TensorProto_DataType_COMPLEX64*/14:
+        case /*onnx::TensorProto_DataType::TensorProto_DataType_COMPLEX128*/15:
+        case /*onnx::TensorProto_DataType::TensorProto_DataType_UNDEFINED*/0:
+        default:
+          resultTypes.push_back(builder.getNoneType());
```

Comment:
Alternative solution is to give attribute to() to getTypeMap.  Suppose the we should have
`static std::vector<int> getTypeMap() {
      retune {mapping_from_protobuf_to_onnx_mlir(to().getSExtValue())}
}`
Can you manually try that? You need to provide mapping_from_protobuf_to_onnx_mlir, which logically is equivalent to what Tian said about reuse the code.
The solution you provide requires less change in gen_onnx_mlir.py. We may come back to this issue after a unified type inference is provided.

Reply:
It seems that we cannot call `to()` inside `getTypeMap` since `getTypeMap` is a static method.

Reply:
You are right.
I am OK with your current implementation even though it has the issue mentioned by @tjingrant .
To accomplish code reuse as Tian mentioned, I am not clear about where that function should be defined. I tried a little bit and ran into link error.

Reply:
I just pushed a commit to make the onnx-to-mlir type translation shareable. Many CMake files are updated to correctly link libraries. If you have any suggestions to reduce CMake updates, please let me know.